crates/sui-indexer-alt/src/pipeline/concurrent/collector.rs

```diff
@@ -0,0 +1,177 @@
+        } else {
+            batch.watermark.push(self.watermark.take(self.values.len()));
+            batch.values.extend(std::mem::take(&mut self.values));
+        }
+    }
+}
+
+impl<H: Handler> From<Indexed<H>> for Pending<H> {
+    fn from(indexed: Indexed<H>) -> Self {
+        Self {
+            watermark: WatermarkPart {
+                watermark: indexed.watermark,
+                batch_rows: indexed.values.len(),
+                total_rows: indexed.values.len(),
+            },
+            values: indexed.values,
+        }
+    }
+}
+
+/// The collector task is responsible for gathering rows into batches which it then sends to a
+/// committer task to write to the database. The task publishes batches in the following
+/// circumstances:
+///
+/// - If `H::BATCH_SIZE` rows are pending, it will immediately schedule a batch to be gathered.
+///
+/// - If after sending one batch there is more data to be sent, it will immediately schedule the
+///   next batch to be gathered (Each batch will contain at most `H::CHUNK_SIZ` rows).
+///
+/// - Otherwise, it will check for any data to write out at a regular interval (controlled by
+///   `config.collect_interval`).
+///
+/// This task will shutdown if canceled via the `cancel` token, or if any of its channels are
+/// closed.
+pub(super) fn collector<H: Handler + 'static>(
+    config: PipelineConfig,
+    mut rx: mpsc::Receiver<Indexed<H>>,
+    tx: mpsc::Sender<Batched<H>>,
+    metrics: Arc<IndexerMetrics>,
+    cancel: CancellationToken,
+) -> JoinHandle<()> {
+    spawn_monitored_task!(async move {
+        // The `poll` interval controls the maximum time to wait between collecting batches,
+        // regardless of number of rows pending.
+        let mut poll = interval(config.collect_interval);
+        poll.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        // Data for checkpoints that haven't been written yet.
+        let mut pending: BTreeMap<u64, Pending<H>> = BTreeMap::new();
+        let mut pending_rows = 0;
+
+        info!(pipeline = H::NAME, "Starting collector");
+
+        loop {
+            tokio::select! {
+                _ = cancel.cancelled() => {
+                    info!(pipeline = H::NAME, "Shutdown received, stopping collector");
+                    break;
+                }
+
+                // Time to create another batch and push it to the committer.
+                _ = poll.tick() => {
+                    let guard = metrics
+                        .collector_gather_latency
+                        .with_label_values(&[H::NAME])
+                        .start_timer();
+
+                    let mut batch = Batched::new();
+                    while !batch.is_full() {
+                        let Some(mut entry) = pending.first_entry() else {
+                            break;
+                        };
+
+                        let indexed = entry.get_mut();
+                        indexed.batch_into(&mut batch);
```

Comment:
Ah I realize that you are basically doing the chunking here in this same loop, instead of in the committer.
I think it would make things easier to reason about if we separate them.
So in this loop we collect processed checkpoint data, and we batch not based on the number of rows but based on the number of checkpoints we processed, because things come in at a fixed checkpoint rate.
When the committer receives the data, it then further chunk it based on its chunk size.

Reply:
Ah I realize that you are basically doing the chunking here in this same loop, instead of in the committer.
I think it would make things easier to reason about if we separate them.
So in this loop we collect processed checkpoint data, and we batch not based on the number of rows but based on the number of checkpoints we processed, because things come in at a fixed checkpoint rate.
When the committer receives the data, it then further chunk it based on its chunk size.

Reply:
What you're describing is roughly how things worked before -- the processor puts out data at checkpoint granularity, and the committer gathered it together and then batched it up for writes on a polling interval.

How we got to here from there is roughly like this:

- I wanted to be able to write to the DB concurrently,
- and because of the bind limits, I need to chunk the writes up before I do that.
- The easiest way to structure a task that does concurrent writes is for it to get work on a stream, and fan it out with `try_for_each_concurrent`.
- This means that I need to have a task that will have that shape (the new `committer`), and another task to fill it (the new `collector`).

Seeing as the `processor` already puts data out on checkpoint boundaries, if we had the collector doing the same it would become redundant, and we still have this need for a separate task to batch and stream chunks.

Reply:
I'm generally not happy with the names `BATCH_SIZE` and `CHUNK_SIZE` -- they are synonymous but the control different things, so suggestions welcome on that. 

- `BATCH_SIZE` controls when we have enough data to justify a DB write, so the factors we use to tune this are latency (lower batch size means better latency) vs one off costs for setting up a DB connection and running the insert (higher batch size means we amortize that cost).
- `CHUNK_SIZE` is easier to explain -- we are limited at the top end by the limits on bind parameters (unless your experiments with `COPY ... FROM` show promise), and also by how long we hold a transaction open for (too long, and the DB doesn't get to do admin tasks, like autovacuum and analysis), but generally we want this to be as large as possible.

Reply:
I have some thoughts on getting rid of CHUNK_SIZE later and will likely experiment on it.
For BATCH_SIZE, since it controls latency, it would be nice if we could transform it into a time based config instead of number of rows. Otherwise I think we will struggle to tune it properly for each pipeline.
If you think that is the right direction, we could add a TODO there but doesn't have to be in this PR.

Reply:
Re: `BATCH_SIZE` being time-based -- we do have both limits: We have a polling interval, and the batch size -- if we accumulate `BATCH_SIZE` many rows before we hit the polling interval then we schedule a write early, otherwise we wait. The idea is that the polling interval makes sure we upperbound latency when we are not under high load, and the `BATCH_SIZE` makes sure we don't unnecessarily create a bottleneck on the DB when we could be letting data out at a steady rate when we are under high load or are backfilling (so we tune it to the database's bandwidth for rows or bytes).

Re: removing `CHUNK_SIZE` -- we can try to do this, but it will require either eliminating the limitations on bind parameters (because that's what it's there to protect against), and being confident that long-running inserts or `COPY`-s don't mess with the database's ability to run admin tasks (like `VACUUM` and `ANALYZE`).

Reply:
For CHUNK_SIZE I was thinking something as simple as default it to 32768 / struct field count, so that it's automatically set for each pipeline without having to tune it, unless someone really want to override it for some reason.
Overall my goal is to rely on as minimum tuning as possible but the default would be good enough in a distributed setting.

Reply:
Got it -- yes, that makes sense. We should still be careful about making sure transactions don't run too long and starve the RDBMS from vacuuming and analyzing, but I suspect that it won't be so much of an issue for concurrent pipelines where this is mainly affected by pruning. Sequential pipelines that modify rows in place may have higher turn-over and so may require further tuning to keep transaction times down, but these are the minority, and they will be handled by a separate pipeline that at the moment does not even have a `CHUNK_SIZE` parameter.